Zone5/Transport/Zone5HTTPClient.swift

```diff
@@ -84,99 +48,151 @@ final public class Zone5HTTPClient {
+		operationQueue.maxConcurrentOperationCount = 1
+		return operationQueue
+	}()
+
+	internal static let downloadQueue: OperationQueue = {
+		let operationQueue = OperationQueue()
+		operationQueue.qualityOfService = .userInitiated
+		operationQueue.maxConcurrentOperationCount = 1
+		return operationQueue
+	}()
+
 	// MARK: Performing requests
 
 	/// The decoder used to convert raw response data into a structure of the expected type.
 	private let decoder = JSONDecoder()
 
-	/// Validates the SDK configuration, and then calls the given `block` function, handling any thrown errors using the given `completion`.
-	///
+	/// A closure that receives the result of a request.
+	internal typealias CompletionHandler<T> = (_ result: Result<T, Zone5.Error>) -> Void
+
+	/// Concurrent queue used to call the completion handlers.
+	private static let completionHandlerQueue = DispatchQueue(label: "Zone5HttpClient.callback.queue", qos: .userInitiated, attributes: .concurrent)
+
+	/// Validates the SDK configuration, and then calls the given `block` closure, handling any thrown errors using the given `completion`.
 	/// - Parameters:
-	///   - completion: Function called with errors that are thrown in the `block`.
-	///   - block: The function containing the work to perform. It receives a strong copy of the parent `Zone5` class, and the configured `baseURL`.
-	///
-	/// - Throws:
-	///	  - Zone5.Error.invalidConfiguration If the Zone5 instance has not been configured with a url
-	///
-	/// - Note:
-	///	  - If neither clientID nor accessToken is configured, the server will respond with an appropriate 401 error
-	private func execute<T>(with completion: @escaping (_ result: Result<T, Zone5.Error>) -> Void, _ block: (_ zone5: Zone5) throws -> PendingRequest) -> PendingRequest? {
+	///   - completion: Closure called with errors that are thrown in the `block`.
+	///   - block: The closure containing the work to perform. It receives a strong copy of the parent `Zone5` class, and the provided `completion`.
+	/// - Throws: `Zone5.Error.invalidConfiguration` if the Zone5 instance has not been configured with a url
+	/// - Note: If neither `clientID` nor `accessToken` is configured, the server will respond with an appropriate 401 error.
+	private func execute<T>(
+		on operationQueue: OperationQueue,
+		with completion: @escaping CompletionHandler<T>,
+		_ block: (_ zone5: Zone5, _ completion: @escaping CompletionHandler<T>) throws -> URLSessionTask
+	) -> PendingRequest? {
 		do {
 			guard let zone5 = zone5, zone5.baseURL != nil else {
 				throw Zone5.Error.invalidConfiguration
 			}
 
-			return try block(zone5)
+			let operation = PendingRequestOperation()
+
+			/// Wrap the provided completion handler in a dispatch call.
+			/// Do this synchronously so that tasks performed afterwards (like deleting files) are handled correctly.
+			let dispatchedCompletion: CompletionHandler<T> = { result in
+				Zone5HTTPClient.completionHandlerQueue.sync {
```

Comment:
Previously this fired completion asynchronously. This means we can't get caught up in a client's code - delaying our own queue of tasks.

If we decide to do this synchronously there is no point in having the DisptachQueue here and the completion can be called in-line.

Maybe we can talk through pros and cons of synchronous vs asynchronous execution here.